### PR TITLE
fix(secrets): grant drain workers access to secrets not yet on the target backend

### PR DESCRIPTION
**Pull request by Gvantsats**
Draining secrets to a new backend failed for unit agents with:
```
unit-hello-0: ERROR juju.worker.dependency "secret-drain-worker" manifold worker returned
unexpected error: executing leadership func: creating policy
"model-secrets-k8s-drain-4fb172-63c320d6-ee91-4ab2-84c8-d71a297bf42f":
Error making API request.

URL: PUT http://192.168.1.213:8200/v1/sys/policies/acl/model-secrets-k8s-drain-4fb172-...
Code: 400. Errors:
* 'policy' parameter not supplied or empty
```
To build the access token for a backend, backendConfigInfo asked the database
which of the accessor's secrets live on that backend. For a drain the backend
is the destination, where the secrets are not yet: the lookup returned nothing,
so no owned secrets were found, so the vault provider produced no policy rules
and vault rejected the empty policy body.

ListGrantedSecretsForBackend now takes forDrain. When set, the backend_uuid
predicate is dropped so revisions are returned from wherever they currently
live, and the secret_value_ref join becomes a LEFT JOIN so revisions held in
the internal backend, which have no value ref row, are returned too. Every
non-drain caller passes false and is unaffected: with the predicate present a
NULL value ref never matches, so the join still behaves as an inner join.

Three consequences of listing revisions across backends are handled:

- Secrets kept in Juju's own database have no external revision ID, so a
  rule cannot name them. The empty IDs are now dropped; the owned set
  still carries the secret IDs, which is all a provider needs.

- Owners hold a manage grant and manage implies view, so the view queries
  also return secrets the accessor owns. These are now dropped from the
  read-only list: vault honours only the most specific matching path, so
  a read-only rule for one revision overrides the owner's wildcard rule
  and removes the write capability the worker needs.

- Kubernetes cannot scope "create" to a named object, so the provider
  pre-creates the secret and grants write access by name. Draining
  defeats this: it pre-creates revision N+1 while the worker rewrites N,
  which is not there yet. Drain tokens may therefore create secrets in
  the model namespace; they are short lived and only issued to the drain
  worker.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
cd tests && ./main.sh -p k8s -c microk8s secrets_k8s test_secret_drain

cd tests && ./main.sh -p k8s -c microk8s secrets_k8s test_user_secret_drain
```

## Links

**Jira card:** [JUJU-10277](https://warthogs.atlassian.net/browse/JUJU-10277)

[JUJU-10277]: https://warthogs.atlassian.net/browse/JUJU-10277?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by gfouillet — CHANGES_REQUESTED** [review]
This is a good analysis yet I don't think wiring down "forDrain" in ListGrantedForBackend is the right approach: for drain causes the backend being ignored, which changes the whole purpose of the function.

It might break other path easily by giving widened privileges.

Prefers a distinct, scoped function, instead of wiring an exception into every layer.

File: domain/secretbackend/service/service.go [hunk 1]
```diff
@@ -308,6 +313,26 @@ func (s *Service) backendConfigInfo(
...
+	// Also drop any read revision for a secret the accessor owns: it is
+	// already covered by the owned rules. Keeping it breaks vault, where an
+	// exact path rule takes precedence over the owner glob rule and so
+	// shadows the write capability needed to update or drain the secret.
+	for id, revisions := range readRevisions {
+		revisions.Remove("")
+		if owned.Contains(id) || revisions.IsEmpty() {
+			delete(readRevisions, id)
+		}
+	}
```
Comment (gfouillet) [inline domain/secretbackend/service/service.go:338]: **todo:** The `readRevisions` dedup-by-owned (`owned.Contains(id)` at line 331) runs unconditionally for all callers of `backendConfigInfo`, not just drain. On the non-drain path, `expandRolesToMatch(RoleView)` returns `[View, Manage]`, so the RoleView query returns manage-role secrets too — meaning owned secrets can now appear in `readRevisions` and are silently dropped.

This is likely a correct latent bug fix (vault exact-path shadowing applies to non-drain callers as well), but it's a behavior change for existing `BackendConfigInfo` callers with no test coverage. The existing regression test `TestBackendConfigInfoOwnedSecretsNotInReadRevisions` uses a non-leader token and only exercises the drain path.

Add a `BackendConfigInfo` test with a **leader token** where the RoleManage and RoleView queries both return the same owned secret, and assert `RestrictedConfig` receives it only in `ownedRevs`, not `readRevs`.

*Cross-ref: builds on lee-van-cleef[bot] S1/S2 — this adds the blast-radius concern that the shared `backendConfigInfo` makes the filtering affect all non-drain callers, not just drain workers.*

File: domain/secretbackend/service/service_test.go [hunk 2]
```diff
@@ -1188,6 +1187,131 @@ func (s *serviceSuite) TestDrainBackendConfigInfo(c *tc.C) {
...
 
+// TestBackendConfigInfoOwnedSecretsNotInReadRevisions is a regression test
+// for the vault drain failure. When the granted secrets are listed across all
+// backends (as the drain worker does), secrets owned by the accessor are also
+// returned by the RoleView queries, so they must be filtered out of the read
+// revisions passed to the provider. Otherwise the vault provider generates an
+// exact path read-only rule for an owned revision, which shadows the glob
+// write rules (an exact path takes precedence over a glob in vault ACLs) and
+// prevents the agent from updating or draining its own secrets.
+func (s *serviceSuite) TestBackendConfigInfoOwnedSecretsNotInReadRevisions(c *tc.C) {
```
Comment (gfouillet) [inline domain/secretbackend/service/service_test.go:1198]: **todo:** No non-drain leader-path test covers the `readRevisions` dedup introduced above. On the leader path, `owners` includes both unit and application, and `expandRolesToMatch(RoleView)` returns `[View, Manage]` — so the RoleView consumer query returns app-managed secrets that are already in `ownedRevisions`. The dedup at `service.go:329-334` drops them, which is correct but untested.

Add a `BackendConfigInfo` (non-drain) test with a leader token (`Check()` returns nil) where the same revision appears in both the RoleManage and RoleView results, and assert it ends up only in `ownedRevs` passed to `RestrictedConfig`.

*Cross-ref: agrees with lee-van-cleef[bot] S2 — the PR's regression test is drain-only and uses a non-leader token, so this specific non-drain scenario remains uncovered.*

Comment (Gvantsats) [reply domain/secretbackend/service/service_test.go:1198]: Added TestBackendConfigInfoLeaderUnitOwnedSecretsNotInReadRevisions covering the non-drain path with a leader token

File: domain/secretbackend/service/service.go (hunk 1 above)
Comment (Gvantsats) [reply domain/secretbackend/service/service.go:338]: Added comments explaining that non-drain path is also affected. Also added TestBackendConfigInfoLeaderUnitOwnedSecretsNotInReadRevisions

**Review by gfouillet — APPROVED** [review]
TA

**Review by iyiguncevik — APPROVED** [review]
